### PR TITLE
Feature/423 cwms flyway

### PR DESCRIPTION
**Pull request by MikeNeilson**
## Problem Description

Fixes #432 

Describe the problem you are trying to solve.

## Solution

Move CWMS CCP Schema elements to flyway style. Created CWMS-Oracle migration provider.
Modified integration-test code to use this.

TODO:
- [x] Allow appropriate bypass options for both CWMS Database and OpenDCS/CCP schema

## how you tested the change

Tested with the automated tests.

## Where the following done:

- [ ] Tests. Check all that apply:
   - [ ] Unit tests created or modified that run during ant test.
   - [ ] Integration tests created or modified that run during integration testing
         (Formerly called regression tests.)
   - [ ] Test procedure descriptions for manual testing
- [ ] Was relevant documentation updated?
- [ ] Were relevant config element (e.g. XML data) updated as appropriate

If you aren't sure leave unchecked and we will help guide you to want needs changing where.

### CODE REVIEW

**Comment by MikeNeilson** [comment]
This is on hold for a while. The TimeSeriesDAO test and CompProc are conflicting with each other in regards to how timeseriesidentifiers are used and updated, add to the cache, etc. It's driving me up the wall and I need to give up for now.

**Comment by MikeNeilson** [comment]
> This is on hold for a while. The TimeSeriesDAO test and CompProc are conflicting with each other in regards to how timeseriesidentifiers are used and updated, add to the cache, etc. It's driving me up the wall and I need to give up for now.

A few general behavior changes were required. Turns out with modifying things to more aggressively use the cached timeseries it was a bit overdone and new time series weren't getting picked up.

It's not ideal in a few places, but seems to sort out the problem and now the CWMS tests are behaving correctly along with the Postgres and XML so it's a solid starting point for future work.

Still need to create the Provider/Test Engine setup for the OpenDCS-Oracle but that will be a separate PR.

@AndrewPWRE if you can create an HDB database similar to how I structured the new CWMS Ready-Database I can help get that organized for HDB.

**Comment by MikeNeilson** [comment]
Okay, I'm done fighting with sonarclouds coverage, there's definitely coverage of the new code, considering it was all done to drive the tests correctly. Will fight that battle in a separate PR.

File: integrationtesting/opendcs-tests/src/test/java/org/opendcs/fixtures/configurations/cwms/CwmsOracleConfiguration.java [hunk 1]
```diff
@@ -32,9 +38,9 @@ public class CwmsOracleConfiguration implements Configuration
 {
     private static Logger log = Logger.getLogger(OpenDCSPGConfiguration.class.getName());
 
-    private static final String CWMS_ORACLE_IMAGE = System.getProperty("opendcs.cwms.oracle.image","docker pull gvenzl/oracle-xe:21.3.0-faststart");
+    private static final String CWMS_ORACLE_IMAGE = System.getProperty("opendcs.cwms.oracle.image","registry-public.hecdev.net/cwms/database-ready-ora-23.5:latest-dev");
```
Comment (ktarbet) [inline integrationtesting/opendcs-tests/src/test/java/org/opendcs/fixtures/configurations/cwms/CwmsOracleConfiguration.java:41]: Is this a 'free' oracle here?

File: java/opendcs/src/main/java/decodes/cwms/CwmsTimeSeriesDb.java [hunk 2]
```diff
@@ -1691,6 +1692,40 @@ public void freeConnection(Connection con)
...
+    public CTimeSeries makeTimeSeries(String tsidStr)
+        throws DbIoException, NoSuchObjectException
+    {
+        try (TimeSeriesDAI timeSeriesDAO = this.makeTimeSeriesDAO())
+        {
+            FailableResult<TimeSeriesIdentifier,TsdbException> tsid = timeSeriesDAO.findTimeSeriesIdentifier(tsidStr, true);
+			if (tsid.isSuccess())
+			{
+				TimeSeriesIdentifier tsId = tsid.getSuccess();
+				// There is an odd situtation that happens were a TimeSeries ID has been load, but never had the storage units attached.
```
Comment (ktarbet) [inline java/opendcs/src/main/java/decodes/cwms/CwmsTimeSeriesDb.java:1715]:
```suggestion
				// There is an odd situation that happens were a TimeSeries ID has been loaded, but never had the storage units attached.
```

File: java/opendcs/src/main/java/decodes/cwms/database/CwmsOracleProvider.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+import decodes.tsdb.ImportComp;
+import decodes.tsdb.TimeSeriesDb;
+import decodes.util.DecodesSettings;
+import ilex.util.EnvExpander;
+
+/**
+ * OpenDCSPgProvider provides support for handling installation and updates of the OpenDCS-Postgres
+ * schema.
+ */
+public class CwmsOracleProvider implements MigrationProvider
```
Comment (ktarbet) [inline java/opendcs/src/main/java/decodes/cwms/database/CwmsOracleProvider.java:33]:
```suggestion
/**
 * CwmsOracleProvider provides support for handling installation and updates of the OpenDCS-CWMS-Oracle
 * schema.
 */
public class CwmsOracleProvider implements MigrationProvider
```

File: java/opendcs/src/main/java/decodes/cwms/database/CwmsOracleProvider.java [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                    assignRole.bind("user",username)
+                              .bind("role",role)
+                              .invoke();
+                }
+            }
+        });*/
+    }
+
+    @Override
+    public List<File> getDecodesData()
```
Comment (ktarbet) [inline java/opendcs/src/main/java/decodes/cwms/database/CwmsOracleProvider.java:119]: The name getDecodesData()   feels like something more than a CwmsOracleProvider, would do, per the class description.  Feels like a higher level OpenDCS setup with Oracle.

File: java/opendcs/src/main/java/decodes/util/ExportTimeSeries.java [hunk 5]
```diff
@@ -151,60 +158,61 @@ protected void runApp()
...
-			else // Should be a time series ID
-			{
-				try
+				else if (TextUtil.startsWithIgnoreCase(outTS, "group:"))
 				{
-					CTimeSeries ts = theDb.makeTimeSeries(outTS);
-					int nvalues = tsDAO.fillTimeSeries(ts, since, until);
-					Logger.instance().info("Read " + nvalues + " values for time series " + outTS);
-					ctss.add(ts);
+					String groupName = outTS.substring(6);
```
Comment (ktarbet) [inline java/opendcs/src/main/java/decodes/util/ExportTimeSeries.java:181]: Suggestion: magic number 6 needs a comment to  explain.

**Review by ktarbet — COMMENTED** [review]
I'm half way through the PR.

Will this flyway work lead to removing dependency on hec-monolith and jooq ?

File: integrationtesting/opendcs-tests/src/test/java/org/opendcs/fixtures/configurations/cwms/CwmsOracleConfiguration.java (hunk 1 above)
Comment (MikeNeilson) [reply integrationtesting/opendcs-tests/src/test/java/org/opendcs/fixtures/configurations/cwms/CwmsOracleConfiguration.java:41]: Yes, the CWMS ready database is based on the oracle free images.

File: java/opendcs/src/main/java/decodes/cwms/database/CwmsOracleProvider.java (hunk 4 above)
Comment (MikeNeilson) [reply java/opendcs/src/main/java/decodes/cwms/database/CwmsOracleProvider.java:119]: Each provider is responsible for declaring it's baseline data. There is some shared data but it was easier at the time to just not worry about it.

File: java/opendcs/src/main/java/decodes/util/ExportTimeSeries.java (hunk 5 above)
Comment (MikeNeilson) [reply java/opendcs/src/main/java/decodes/util/ExportTimeSeries.java:181]: hmm, yeah. length of the word "group:" should define group as a constant and use length.

**Comment by MikeNeilson** [comment]
> I'm half way through the PR.
> 
> Will this flyway work lead to removing dependency on hec-monolith and jooq ?

By itself no, but since the CWMS Operations are under test easily the work is easier.

Also we currently depend on monolith because of hec-cwms-ratings so getting rid of monolith will take a bit longer.

File: java/opendcs/src/main/resources/db/CWMS-Oracle/callbacks/afterMigrate__setPerms.sql
Comment (ktarbet) [inline java/opendcs/src/main/resources/db/CWMS-Oracle/callbacks/afterMigrate__setPerms.sql:1]: The comments mention: ccp_users_w and ccp_users_p

However, the code in this file operations on CCP_USERS in all cases.

File: java/opendcs/src/main/resources/db/CWMS-Oracle/packages/R__cwms_vpd.sql
Comment (ktarbet) [inline java/opendcs/src/main/resources/db/CWMS-Oracle/packages/R__cwms_vpd.sql:1]: fyi - I don't understand  some of this code such as :

```sql
 l_pred := '1 = 1'
```

File: java/opendcs/src/main/resources/db/CWMS-Oracle/schema/V6.8__opendcs.sql
Comment (ktarbet) [inline java/opendcs/src/main/resources/db/CWMS-Oracle/schema/V6.8__opendcs.sql:1]: Is this the first possible version?    The commands seem to assume the tables for example don't exist.

**Review by ktarbet — COMMENTED** [review]
I've been through all these files.   I don't have experience to understand all the SQL .

How can we test updating using this flyway strategy?

What about installations that only use the decodes portion of OpenDCS?  Perhaps they have the CCP tables but don't know about them?

File: java/opendcs/src/main/java/decodes/cwms/database/CwmsOracleProvider.java [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+import decodes.tsdb.ImportComp;
+import decodes.tsdb.TimeSeriesDb;
+import decodes.util.DecodesSettings;
+import ilex.util.EnvExpander;
+
+/**
+ * CwmsOracleProvider provides support for handling installation and updates of the OpenDCS-CWMS-Oracle
+ * schema.
+ */
+public class CwmsOracleProvider implements MigrationProvider
```
[the remark this answers is not in the document]
Comment (MikeNeilson) [reply java/opendcs/src/main/java/decodes/cwms/database/CwmsOracleProvider.java:33]: The name matches existing implementations. While part of an installer, it providers information to the installer.

**Comment by MikeNeilson** [comment]
> I've been through all these files. I don't have experience to understand all the SQL .
> 
> How can we test updating using this flyway strategy?
> 
> What about installations that only use the decodes portion of OpenDCS? Perhaps they have the CCP tables but don't know about them?

There was never a distinction for the CWMS installs. It's either all or nothing.

For testing flyway, the integration test includes two versioned scripts, one is applied after the other.

If you mean testing on an existing installation. Not possible, we have to perform some manual steps that aren't exposed yet to "baseline" to flyway, since the schema is already installed. That's a later project where this is just focused on initial setup and making sure the tests can be run.

File: java/opendcs/src/main/resources/db/CWMS-Oracle/packages/R__cwms_vpd.sql
Comment (MikeNeilson) [reply java/opendcs/src/main/resources/db/CWMS-Oracle/packages/R__cwms_vpd.sql:1]: That's part of the VPD mechanism. You set a statement that gets added to the where clause in the background that is true for the rows that should be returned.

For our usage a simple 1=1 or 1=0 is sufficient. A more complex system could include columns from the query itself.

File: java/opendcs/src/main/resources/db/CWMS-Oracle/callbacks/afterMigrate__setPerms.sql
Comment (MikeNeilson) [reply java/opendcs/src/main/resources/db/CWMS-Oracle/callbacks/afterMigrate__setPerms.sql:1]: Yeah, there wasn't never much of a distinction.

**Review by ktarbet — DISMISSED** [review]

**Review by ktarbet — APPROVED** [review]
Github shows no new changes (in the Files Changed tab) since my last review.

**Pull request merged**